contrib/aws/aws-sdk-go-v2/aws/aws.go

```diff
@@ -101,6 +126,228 @@ func (mw *traceMiddleware) startTraceMiddleware(stack *middleware.Stack) error {
 	}), middleware.After)
 }
 
+func extractResourceNameFromParams(requestInput middleware.InitializeInput, awsService string) (string, string, error) {
+	var resourceNameKey, resourceNameValue string
+
+	switch awsService {
+	case "SQS":
+		resourceNameKey = tagQueueName
+		resourceNameValue = extractQueueName(requestInput)
+	case "S3":
+		resourceNameKey = tagBucketName
+		resourceNameValue = extractBucketName(requestInput)
+	case "SNS":
+		resourceNameKey = tagTopicName
+		resourceNameValue = extractTopicName(requestInput)
+	case "DynamoDB":
+		resourceNameKey = tagTableName
+		resourceNameValue = extractTableName(requestInput)
+	case "Kinesis":
+		resourceNameKey = tagStreamName
+		resourceNameValue = extractStreamName(requestInput)
+	case "EventBridge":
+		resourceNameKey = tagRuleName
+		resourceNameValue = extractRuleName(requestInput)
+	case "SFN":
+		resourceNameKey = tagStateMachineName
+		resourceNameValue = extractStateMachineName(requestInput)
+	default:
+		return "", "", fmt.Errorf("attemped to extract ResourceNameFromParams of an unsupported AWS service: %s", awsService)
+	}
+
+	return resourceNameKey, resourceNameValue, nil
+}
+
+func extractQueueName(requestInput middleware.InitializeInput) string {
+	var queueURL *string
+	switch params := requestInput.Parameters.(type) {
+	case *sqs.SendMessageInput:
+		queueURL = params.QueueUrl
+	case *sqs.DeleteMessageInput:
+		queueURL = params.QueueUrl
+	case *sqs.DeleteMessageBatchInput:
+		queueURL = params.QueueUrl
+	case *sqs.ReceiveMessageInput:
+		queueURL = params.QueueUrl
+	case *sqs.SendMessageBatchInput:
+		queueURL = params.QueueUrl
+	}
+	if queueURL == nil {
+		return ""
+	}
+	parts := strings.Split(*queueURL, "/")
+	return parts[len(parts)-1]
+}
+
+func extractBucketName(requestInput middleware.InitializeInput) string {
+	var bucket *string
+	switch params := requestInput.Parameters.(type) {
+	case *s3.ListObjectsInput:
+		bucket = params.Bucket
+	case *s3.ListObjectsV2Input:
+		bucket = params.Bucket
+	case *s3.PutObjectInput:
+		bucket = params.Bucket
+	case *s3.GetObjectInput:
+		bucket = params.Bucket
+	case *s3.DeleteObjectInput:
+		bucket = params.Bucket
+	case *s3.DeleteObjectsInput:
+		bucket = params.Bucket
+	}
+	if bucket == nil {
+		return ""
+	}
+	return *bucket
```

Comment:
```suggestion
	switch params := requestInput.Parameters.(type) {
	case *s3.ListObjectsInput, s3.ListObjectsV2Input, *s3.PutObjectInput, *s3.GetObjectInput, *s3.DeleteObjectInput, *s3.DeleteObjectsInput:
		return *params.Bucket
	default:
		return ""
	}
```

Reply:
Here and throughout, where feasible.

Reply:
The suggested change you gave doesn't work, I end up getting params.Bucket undefined (type interface{} has no field or method Bucket)compilerMissingFieldOrMethod

Unfortunately, it doesn't seem possible to group cases with different types in a type switch and access their specific fields.

It seems go does not allow you to access specific fields when grouping different types because, in this situation, the type information is lost. As a result, you cannot group the cases while still accessing the Bucket field for each type in the switch statement.

Reply:
Ah bummer. Okay 👍